pep-0708.rst

```diff
@@ -0,0 +1,849 @@
+someone is able to publish package ``B`` to repository ``Y`` under the same name.
+
+A specific example of this is the recent case where the PyTorch project had an
+internal project named ``torchtriton`` which was only ever intended to be
+installed from their repositories located at ``https://download.pytorch.org/``,
+but that repository was designed to be used in conjunction with PyPI, and
+the name of ``torchtriton`` was not claimed on PyPI, which allowed the attacker
+to use that name and publish a malicious version.
+
+This PEP proposes an extension to the :ref:`Simple Repository API
+<packaging:simple-repository-api>` that would enable
+installers to determine when it is safe to install a given package that appears
+in multiple repositories, allowing them to be stricter in the cases where they
+do not know if it is safe.
+
+
+Motivation
+===========
+
+Dependency Confusion attacks have long been possible, but they've recently
+gained press with public examples of cases where these attacks were successfully
+executed.
+
+There are a number of ways to mitigate against these attacks today, but they all
+require that the end user go out of their way to protect themselves, rather than
+being protected by default. This means that for the vast bulk of users, they are
+likely to remain vulnerable, even if they are ultimately aware of these types of
+attacks.
+
+Ultimately the underlying cause of these attacks come from the fact that there
+is no globally unique namespace that all Python package names come from.
+Instead, each repository is its own distinct namespace, and when given an
+"abstract" name such as ``spam`` to install, an installer has to implicitly turn
+that into a "concrete" name such as ``pypi.org:spam`` or ``example.com:spam``.
+Currently the standard behavior in Python installation tools is to implicitly
+flatten these multiple namespaces into one that contains the files from all
+namespaces.
+
+This implicit assumption that collapsing the namespaces is what was expected,
+means that that when it was not (such as in the ``torchtriton`` case) dependency
+confusion attacks become possible.
+
+What makes this particularly tricky is two competing things:
+
+1. Collapsing the namespaces is exactly the behavior that you want at times;
+   for instance, the `Piwheels project <https://www.piwheels.org/>`_
+   does not manage a namespace of its own,
+   but rather exists to supplment the namespace from PyPI with additional files.
+2. Not collapsing the namespaces is the behavior that you want for instances
+   like ``torchtriton`` where the name on PyPI and the name on PyTorch's
+   repositories were fundamentally different things.
+
+So an installer needs some mechanism by which to determine when it should
+flatten the namespaces of multiple repositories and when it should not, rather
+than having to choose between always flattening and never flattening.
+
+This functionaly could be pushed directly into the end user, since ultimately
```

Comment:
Found via `make spellcheck` (and including CAM's suggestion):
```suggestion
This functionality could be pushed directly to the end user, since ultimately
```